src/Microsoft.Management.Deployment/PackageManager.cpp

```diff
@@ -52,6 +57,95 @@ namespace winrt::Microsoft::Management::Deployment::implementation
+        {
+            auto sourceList = ::AppInstaller::Repository::Source::GetCurrentSources();
+
+            std::string sourceType = type;
+
+            // [NOTE:] If the source type is not specified, the default source type will be used for validation.In cases where the source type is empty,
+            // it remains unassigned until the add operation, at which point it is assigned.Without this default assignment, an empty string could be
+            // compared to the default type, potentially allowing different source names with the same URI to be seen as unique.
+            // To avoid this, assign the default source type prior to comparison.
+            if (sourceType.empty())
+            {
+                // This method of obtaining the default source type is slightly expensive as it requires creating a SourceFactory object
+                // and fetching the type name.Nonetheless, it future-proofs the code against any changes in the SourceFactory's default type.
+                sourceType = ::AppInstaller::Repository::Source::GetDefaultSourceType();
+            }
+
+            for (const auto& source : sourceList)
+            {
+                THROW_HR_IF(APPINSTALLER_CLI_ERROR_SOURCE_NAME_ALREADY_EXISTS, ::AppInstaller::Utility::ICUCaseInsensitiveEquals(source.Name, name));
+
+                bool sourceUriAlreadyExists = !source.Arg.empty() && source.Arg == sourceUri && source.Type == sourceType;
+                THROW_HR_IF(APPINSTALLER_CLI_ERROR_SOURCE_ARG_ALREADY_EXISTS, sourceUriAlreadyExists);
+            }
+        }
+
+        ::AppInstaller::Repository::Source CreateSourceFromOptions(const winrt::Microsoft::Management::Deployment::AddPackageCatalogOptions& options)
+        {
+            std::string name = winrt::to_string(options.Name());
+            std::string type = winrt::to_string(options.Type());
+            std::string sourceUri = winrt::to_string(options.SourceUri());
+
+            AppInstaller::Repository::SourceTrustLevel trustLevel = AppInstaller::Repository::SourceTrustLevel::None;
+            if (options.TrustLevel() == winrt::Microsoft::Management::Deployment::PackageCatalogTrustLevel::Trusted)
+            {
+                trustLevel = AppInstaller::Repository::SourceTrustLevel::Trusted;
+            }
+
+            CheckForDuplicateSource(name, type, sourceUri);
+
+            ::AppInstaller::Repository::Source source = ::AppInstaller::Repository::Source{ name, sourceUri, type, trustLevel, options.Explicit() };
+
+            std::string customHeader = winrt::to_string(options.CustomHeader());
+            if (!customHeader.empty())
+            {
+                source.SetCustomHeader(customHeader);
+            }
+
+            auto sourceInfo = source.GetInformation();
+
+            if (sourceInfo.Authentication.Type == ::AppInstaller::Authentication::AuthenticationType::Unknown)

```

Comment:
@yao-msft - This check seems particularly relevant to the REST Source scenario. Do you think it's needed for the Add operations? I'm trying to confirm whether this is a redundancy check or if it becomes relevant when the caller invokes the PackageManagerReference.Connect API call. If it's redundant, I'll go ahead and remove this check and the associated error result, as we no longer encounter that scenario during the add call.

Reply:
I think we can keep it. If a source is not supported by the client, we'd better return an error instead of adding it because all subsequent connect calls will fail in the future.